modules/NuGet.Tasks/ApplyNuGetPolicies.cs

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using Newtonsoft.Json;
+using NuGet.Tasks.Policies;
+using NuGet.Tasks.ProjectModel;
+using NuGet.Tasks.Utilties;
+
+namespace NuGet.Tasks
+{
+    /// <summary>
+    /// Applies policies to a repository on how NuGet can be used, such as restricting which PackageReference's are allowed,
+    /// which versions of packages can be used, which feeds are available, etc.
+    /// </summary>
+    public class ApplyNuGetPolicies : Microsoft.Build.Utilities.Task, ICancelableTask
+    {
+        private readonly CancellationTokenSource _cts;
+        private readonly MSBuildPolicyFactory _policyFactory;
+
+        public ApplyNuGetPolicies()
+        {
+            _policyFactory = new MSBuildPolicyFactory();
+            _cts = new CancellationTokenSource();
+        }
+
+        [Required]
+        public string SolutionDirectory { get; set; }
+
+        /// <summary>
+        /// The policies to be applied to the repository.
+        /// </summary>
+        [Required]
+        public ITaskItem[] Policies { get; set; }
+
+        /// <summary>
+        /// The MSBuild projects or solutions to which the policy should be applied.
+        /// </summary>
+        [Required]
+        public ITaskItem[] Projects { get; set; }
+
+        /// <summary>
+        /// e.g. "Configuration=Debug;BuildNumber=1234"
+        /// </summary>
+        public string Properties { get; set; }
```

Comment:
'Properties' is a little vague. I'm going to rename to 'ProjectProperties'. This are properties meant to be applied to projects and solutions when they are loaded. This task executes a design-time build in order to examine target frameworks, package references, and a few other details.